internal/cloudsql/instance.go

```diff
@@ -293,9 +293,21 @@ func (i *Instance) scheduleRefresh(d time.Duration) *refreshOperation {
 		res.md, res.tlsCfg, res.expiry, res.err = i.r.performRefresh(i.ctx, i.connName, i.key, i.RefreshCfg.UseIAMAuthN)
 		close(res.ready)
 
+		select {
+		case <-i.ctx.Done():
+			// instance has been closed, don't schedule anything
+			return
+		default:
+		}
+
 		// Once the refresh is complete, update "current" with working result and schedule a new refresh
 		i.resultGuard.Lock()
 		defer i.resultGuard.Unlock()
+
+		// schedule the next refresh in the future
+		t := refreshDuration(time.Now(), i.cur.expiry)
+		i.next = i.scheduleRefresh(t)
```

Comment:
I think this refresh needs to stay put, otherwise there will be 2 refreshes scheduled in the error case (that next clause has a return that prevents these statements from being reached)

Reply:
Ah, yep thanks for pointing that out. Updated